### DISCUSSION TITLE
How to properly install wheel file built from source?

### DISCUSSION THREAD

**Discussion by mattroos** [microsoft/onnxruntime#7603] [Other Q&A]
I know this is really a python/pip question, but hopefully someone can help...

I've built onnxruntime from source,
```
 ./build.sh --config Release --build_shared_lib --parallel --build_wheel
```
and then installed the wheel file...
```
pip install ./build/Linux/Release/dist/onnxruntime_gpu_tensorrt-1.7.0-cp36-cp36m-linux_x86_64.whl
```
But it fails to import into python.
```
>>> import onnxruntime
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'onnxruntime'
```
If I then do a ```pip uninstall```, it indicates that the installed files were in the expected directories (in the virtual python environment in which I did the ```pip install```).
```
mroos@synapse:~$ pip uninstall ./build/Linux/Release/dist/onnxruntime_gpu_tensorrt-1.7.0-cp36-cp36m-linux_x86_64.whl 
Found existing installation: onnxruntime-gpu-tensorrt 1.7.0
Uninstalling onnxruntime-gpu-tensorrt-1.7.0:
  Would remove:
    /home/mroos/python_envs/env_p3.6.9_tf2.2/bin/onnxruntime_test
    /home/mroos/python_envs/env_p3.6.9_tf2.2/lib/python3.6/site-packages/onnxruntime/*
    /home/mroos/python_envs/env_p3.6.9_tf2.2/lib/python3.6/site-packages/onnxruntime_gpu_tensorrt-1.7.0.dist-info/*
Proceed (y/n)? y
  Successfully uninstalled onnxruntime-gpu-tensorrt-1.7.0
```
What am I doing wrong? Why can't python find and import the module?

**Comment by ytaous** [comment]
hi, @mattroos - usually I do:

1 cd to dist dir
2. pip install onnx onnxruntime_gpu_tensorrt-1.7.0-cp36-cp36m-linux_x86_64.whl --force
3. verify with import again

**Reply by mattroos** [reply]
Thanks, @ytaous. What does the ```onnx``` before ```onnxruntime_gpu_tensorrt-1.7.0-cp36-cp36m-linux_x86_64.whl``` tell ```pip``` to do? And is ```--force``` a shorthand for ```--force_reinstall```? I did a brief browse of the pip documentation, but the answers weren't forthcoming.

In any case, I followed your guidance and it worked, although when importing the module I then got errors like this one:
```
>>> import onnxruntime
/home/mroos/python_envs/env_p3.6.9_tf2.2/lib/python3.6/site-packages/onnxruntime/capi/_pybind_state.py:14: UserWarning: Cannot load onnxruntime.capi. Error: 'libcudart.so.10.2: cannot open shared object file: No such file or directory'.
  warnings.warn("Cannot load onnxruntime.capi. Error: '{0}'.".format(str(e)))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/mroos/python_envs/env_p3.6.9_tf2.2/lib/python3.6/site-packages/onnxruntime/__init__.py", line 24, in <module>
    from onnxruntime.capi._pybind_state import get_all_providers, get_available_providers, get_device, set_seed, \
ImportError: cannot import name 'get_all_providers'
```
I tried creating some symbolic links from version 10 of cublas and cudart files to my version 11 files, but it seemed to be going down a rabbit hole. So, I deleted by entire build directory, made sure I only have CUDA 11.3 available (as best I could), and built again. Now I'm getting this test error after the build:
```
2021-05-07 09:19:27,486 util.run [INFO] - Running subprocess in '/home/mroos/Code/onnxruntime/build/Linux/Release'
['/home/mroos/python_envs/env_p3.6.9_tf2.2/bin/python3', '/home/mroos/Code/onnxruntime/onnxruntime/test/onnx/gen_test_models.py', '--output_dir', 'test_models']
Traceback (most recent call last):
  File "/home/mroos/Code/onnxruntime/onnxruntime/test/onnx/gen_test_models.py", line 154, in <module>
    test_size(args.output_dir)
  File "/home/mroos/Code/onnxruntime/onnxruntime/test/onnx/gen_test_models.py", line 143, in test_size
    os.path.join(output_dir, 'test_size_float'))
  File "/home/mroos/Code/onnxruntime/onnxruntime/test/onnx/gen_test_models.py", line 81, in generate_size_op_test
    final_model = onnx.utils.polish_model(model_def)
AttributeError: module 'onnx.utils' has no attribute 'polish_model'
Traceback (most recent call last):
  File "/home/mroos/Code/onnxruntime/tools/ci_build/build.py", line 1986, in <module>
    sys.exit(main())
  File "/home/mroos/Code/onnxruntime/tools/ci_build/build.py", line 1924, in main
    run_onnxruntime_tests(args, source_dir, ctest_path, build_dir, configs)
  File "/home/mroos/Code/onnxruntime/tools/ci_build/build.py", line 1401, in run_onnxruntime_tests
    '--output_dir', 'test_models'], cwd=cwd)
  File "/home/mroos/Code/onnxruntime/tools/ci_build/build.py", line 528, in run_subprocess
    return run(*args, cwd=cwd, capture_stdout=capture_stdout, shell=shell, env=my_env)
  File "/home/mroos/Code/onnxruntime/tools/python/util/run.py", line 44, in run
    env=env, shell=shell)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/home/mroos/python_envs/env_p3.6.9_tf2.2/bin/python3', '/home/mroos/Code/onnxruntime/onnxruntime/test/onnx/gen_test_models.py', '--output_dir', 'test_models']' returned non-zero exit status 1.
```
I started python and imported onnx.utils, and sure enough, it has no attribute ```polish_model``` attribute. I have the latest onnx release installed, version 1.9.0. What might be the problem?

**Reply by pranav-prakash** [reply]
@mattroos `polish_models` was removed from the `gen_test_models` script in https://github.com/microsoft/onnxruntime/commit/b22e60bd442cdd9a96138a5705418a76ef46e467#diff-7ffbbbc3394c2ca2f5c48cd03d84863347becf98ded4d6cd22168ad30716abdf in March.

**Reply by mattroos** [reply]
Thanks, @pranav-prakash. I'm trying to make understand what that implies, and what my next steps should be. I think it means this, please correct me if I'm wrong...

The command ```pip install onnx onnxruntime_gpu_tensorrt-1.7.0-cp36-cp36m-linux_x86_64.whl --force``` directed pip install install onnx, and then to install the onnxruntime wheel. In doing so, it installed a recent version of onnx that does not include ```polish_models```. Thus, when I then tried to rebuild onnxruntime, the build instructions tried to used ```polish_models```, which was not available. This halted the build, and the wheel was not created.

**But what is the best approach to resolving this?** Installed an older version of onnx? Change something in the build instructions for onnxruntime?

**Reply by pranav-prakash** [reply]
@mattroos I meant that it's weird the script is still calling into `polish_models` given that that linked commit to onnxruntime removed the call at the site your stack trace seems to indicate. I'm not sure how you obtained the `onnxruntime_gpu_tensorrt-1.7.0` wheel (whether you build from HEAD or built a specific version) but build `1.7.0` (at least according to their release changelog) was from March 2 and didn't include that linked commit. So I'd suggest building the current HEAD, assuming you haven't already been doing that (if you have, then it's more of a mystery)

**Reply by jywu-mysoft** [reply]
as a workaround, you can add --update --test to your build command to skip tests. this way the wheel can get built successfully.

**Reply by mattroos** [reply]
@pranav-prakash, I would have bet my life that I did a fresh ```git pull``` just before the build, but I guess I didn't. I now did a new pull, which fixed the ```polish_model``` error as expected. Then the build and test ran to completion, with no glaring problems. The ```pip install``` gave no errors as well. But, I now get this error when importing into python:
```
Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!
```
Any guess?

PS - Thanks to @jywu-msft for that kind suggestion.

**Reply by pranav-prakash** [reply]
@mattroos Something seems to be going wrong with one of the native shared libraries that ORT python package depends on. I'm a bit less familiar with this territory, but you can try setting the `LD_DEBUG=all` environment variable to see if it prints any more info.

**Reply by mattroos** [reply]
@pranav-prakash, that produced a lot of output after the ```import```, the end of which is this:
```
15198:	checking for version 'GCC_3.0' in file /lib/x86_64-linux-gnu/libgcc_s.so.1 [0] required by file /home/mroos/python_envs/env_torch/lib/python3.6/site-packages/onnxruntime/capi/onnxruntime_pybind11_state.cpython-36m-x86_64-linux-gnu.so [0]
15198:	checking for version 'GLIBC_2.3' in file /lib64/ld-linux-x86-64.so.2 [0] required by file /home/mroos/python_envs/env_torch/lib/python3.6/site-packages/onnxruntime/capi/onnxruntime_pybind11_state.cpython-36m-x86_64-linux-gnu.so [0]
Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!
```
Is this implying ORT requires glibc 2.3? 2.3 or lower? 2.3 or higher?
It appears I have 2.27 installed, which I believe is the default of Ubuntu 18.04.
```
mroos@synapse:~$ ldd --version
ldd (Ubuntu GLIBC 2.27-3ubuntu1.4) 2.27
```
I'm not sure how to upgrade it, and if I do so, whether or not it will break something else.

**Reply by pranav-prakash** [reply]
@mattroos
If I'm reading the `ld` source correctly (https://code.woboq.org/userspace/glibc/elf/dl-version.c.html) it does seem to indicate that pybind was linked against glibc 2.3 (and thus requires 2.3 or higher). This seems a bit strange though, since if you only have glibc 2.27 on your system then I'm not sure how your compiler linked against a newer glibc version (just to double check though, this was compiled on the same system that you're running it on, right?)

**Reply by mattroos** [reply]
Yes, compiled and used on the same machine, same python environment. I wrote a short C program to confirm that gcc was linking to glibc 2.27...
```
mroos@synapse:scratch$ cat temp_glibc_version.c 
#include <stdio.h>
#include <stdlib.h>
#include <gnu/libc-version.h>

int main(int argc, char *argv[]) {
  printf("GNU libc version: %s\n", gnu_get_libc_version());
  exit(EXIT_SUCCESS);
}
```
```
mroos@synapse:scratch$ gcc temp_glibc_version.c -o temp_glibc
mroos@synapse:scratch$ ./temp_glibc 
GNU libc version: 2.27
```
I'm not sure what to make of this, from the LD_DEBUG output:
```
checking for version 'GCC_3.0' in file /lib/x86_64-linux-gnu/libgcc_s.so.1 [0] required by file /home/mroos/python_envs/env_torch/lib/python3.6/site-packages/onnxruntime/capi/onnxruntime_pybind11_state.cpython-36m-x86_64-linux-gnu.so [0]
```
gcc 3.0? So old. My system has gcc 7.5.0, and I don't see any other versions in ```/usr/bin```.

**Reply by snnn** [reply]
Hi @mattroos , when you build onnxruntime from source, you should install onnx from source too.  Otherwise you may see strange test failures like the one in gen_test_models.py.

**Reply by snnn** [reply]
The "Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!" issue is an old bug coming from the patchelf used by auditwheel. You may ignore the warning. Python community plans to replace the tool: https://github.com/pypa/auditwheel/issues/174 . I'm sorry we can't do much on that.

**Reply by mattroos** [reply]
Thanks for the tips and info, @snnn!

**Reply by mattroos** [reply]
Oh, I had forgotten until touching my code again... @snnn, I can't ignore the warning because it's not a warning. When I do the ```import onnxruntime```, I get that message and then my python session exits/ends/crashes.

**Reply by snnn** [reply]
If you built onnxruntime as " ./build.sh --config Release --build_shared_lib --parallel --build_wheel", then the result package should not have tensorrt. Did I miss something?
Are you using Ubuntu 18.04? If you are not building onnxruntime inside of a manylinux docker image(which is centos7), then you should not hit the auditwheel issue. If auditwheel didn't call patchelf, then you should not see the ld.so warning. So it looks strange to me.

**Reply by mattroos** [reply]
@snnn, my apologies! I see now that I ran off course at some point. I've been trying/wanting to build a version that uses my NVIDIA RTX2080 gpu, but as some point I inadvertently started using the wrong build command line options (erroneous cut/paste). The built version that currently produces the error when importing onnxruntime is a cpu version (```onnxruntime-1.7.0-cp36-cp36m-linux_x86_64.whl```).

I'm going to start from scratch, and try to again build onnxruntime with tensorrt. Will I need to build onnx from source in that case as well?

[EDIT: Addition]
By the way, my end goal is convert a pytorch model to onnx and run it on a Jetson AGX Xavier. Doing this on my RTX gpu is just an intermediate step, and test to see how differently the onnx model runs and performs. Any guidance or warnings about running tensorrt-onnx models on a Jetson are welcome!

**Reply by snnn** [reply]
For the ld.so issue I think we can start with the CPU case, it's simpler.  When you build the wheel ,did you see logs like:
```
repairing [..something...] for manylinux1
```
?

**Reply by mattroos** [reply]
Unfortunately, @snnn, I deleted the old build directory before reading your response. But I'm now trying to follow your overall guidance, by building both onnx and onnxruntime from source, and sticking with a CPU for now. I did fresh pulls of onnx and onnxruntime. I then built and installed onnx, with no warnings or errors. Finally, I built onnxruntime, ```./build.sh --config Release --build_shared_lib --parallel --enable_pybind --build_wheel``` and received many test errors, including some like the one below, which halted the testing.
```
======================================================================
ERROR: test_quantize_conv_without_bias (test_qdq.TestQDQFormatConvClip)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mroos/Code/onnxruntime/build/Linux/Release/quantization/test_qdq.py", line 184, in test_quantize_conv_without_bias
    self.verify(False) # per_channel:False
  File "/home/mroos/Code/onnxruntime/build/Linux/Release/quantization/test_qdq.py", line 163, in verify
    reduce_range = per_channel
  File "/home/mroos/Code/onnxruntime/build/Linux/Release/onnxruntime/quantization/quantize.py", line 195, in quantize_static
    model = load_model(Path(model_input), optimize_model)
  File "/home/mroos/Code/onnxruntime/build/Linux/Release/onnxruntime/quantization/quantize.py", line 48, in load_model
    onnx_model = ONNXModel(optimize_model(Path(model_path)))
  File "/home/mroos/Code/onnxruntime/build/Linux/Release/onnxruntime/quantization/quantize.py", line 40, in optimize_model
    _ = InferenceSession(model_path.as_posix(), sess_option, providers=['CPUExecutionProvider'])
  File "/home/mroos/Code/onnxruntime/build/Linux/Release/onnxruntime/capi/onnxruntime_inference_collection.py", line 283, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/mroos/Code/onnxruntime/build/Linux/Release/onnxruntime/capi/onnxruntime_inference_collection.py", line 310, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.Fail: [ONNXRuntimeError] : 1 : FAIL : Load model from conv_clip_fp32.False.onnx failed:/home/mroos/Code/onnxruntime/onnxruntime/core/graph/model.cc:111 onnxruntime::Model::Model(onnx::ModelProto&&, const PathString&, const IOnnxRuntimeOpSchemaRegistryList*, const onnxruntime::logging::Logger&) Unknown model file format version.
```

**Reply by snnn** [reply]
@yufenglee Any idea about the QDQ test failure?

**Reply by yufenglee** [reply]
The error says "Unknown model file format version.". ONNX has a [IR update](https://github.com/onnx/onnx/blob/8d15ccc183033f44e7cbcda480bfa0b3d8070efd/onnx/onnx.proto#L97) recently, but it is not pulled and supported in onnxruntime. You can resolve this issue by rebuilding and reinstall your onnx from the [stable release branch](https://github.com/onnx/onnx/commits/rel-1.9.1) .

**Reply by yufenglee** [reply]
I have a [PR ](https://github.com/microsoft/onnxruntime/pull/7747)to avoid this kind of failure.

**Reply by mattroos** [reply]
Thanks, @yufenglee! I got pulled onto something else for a few days but will git a go after that.

**Reply by mattroos** [reply]
Hi, @snnn, @yufenglee, @pranav-prakash, and @ytaous! I'm back on this effort after a hiatus.

After building and installing onnx 1.9.0 and the most resent onnxruntime pull, I'm able to import the CPU version of both into python. I'm now trying to build the TensorRT version. I initially tried with the recent tensorrt version 8.0.0.3. Here is my build command:
```bash
./build.sh \
--config Release \
--build_shared_lib \
--parallel \
--build_wheel \
--use_tensorrt \
--tensorrt_home /usr/src/tensorrt \
--cudnn_home /usr/lib/x86_64-linux-gnu \
--cuda_home /usr/local/cuda
```
I get this missing library error:
```bash
-- Found TensorRT headers at /usr/include/x86_64-linux-gnu
-- Find TensorRT libs at /usr/lib/x86_64-linux-gnu/libnvinfer.so;/usr/lib/x86_64-linux-gnu/libnvinfer_plugin.so;TENSORRT_LIBRARY_MYELIN-NOTFOUND
-- Could NOT find TENSORRT (missing: TENSORRT_LIBRARY) 
ERRORCannot find TensorRT library.
```
Do I need a lower version of TensorRT? [Here](https://www.onnxruntime.ai/docs/how-to/build/eps.html#tensorrt) they state that onnxruntime was built and tested with tensorrt 7.2.2 and 7.2.3.
